File: majorProject/src/main/java/com/aocc/framework/Viewport.java
package com.aocc.framework;

import android.graphics.Rect;

/**
 * Maps between virtual world coordinates and the letterboxed on-screen area.
 */
public class Viewport {

    private float scale = 1f;
    private float offsetX;
    private float offsetY;
    private int viewWidth;
    private int viewHeight;

    public void update(int viewWidth, int viewHeight) {
        if (viewWidth <= 0 || viewHeight <= 0) {
            return;
        }

        this.viewWidth = viewWidth;
        this.viewHeight = viewHeight;

        float scaleX = (float) viewWidth / GameConstants.WORLD_WIDTH;
        float scaleY = (float) viewHeight / GameConstants.WORLD_HEIGHT;
        scale = Math.min(scaleX, scaleY);

        float contentWidth = GameConstants.WORLD_WIDTH * scale;
        float contentHeight = GameConstants.WORLD_HEIGHT * scale;
        offsetX = (viewWidth - contentWidth) * 0.5f;
        offsetY = (viewHeight - contentHeight) * 0.5f;
    }

    public int screenToWorldX(float screenX) {
        return Math.round((screenX - offsetX) / scale);
    }

    public int screenToWorldY(float screenY) {
        return Math.round((screenY - offsetY) / scale);
    }

    public Rect getLetterboxDestRect() {
        Rect rect = new Rect();
        rect.left = Math.round(offsetX);
        rect.top = Math.round(offsetY);
        rect.right = Math.round(offsetX + GameConstants.WORLD_WIDTH * scale);
        rect.bottom = Math.round(offsetY + GameConstants.WORLD_HEIGHT * scale);
        return rect;
    }

    public float getScale() {
        return scale;
    }

    public int getViewWidth() {
        return viewWidth;
    }

    public int getViewHeight() {
        return viewHeight;
    }
}

File: majorProject/src/main/kotlin/com/aocc/framework/Viewport.kt
package com.aocc.framework

import android.graphics.Rect

/**
 * Maps between virtual world coordinates and the letterboxed on-screen area.
 */
class Viewport {

    private var scale: Float = 1f
    private var offsetX: Float = 0f
    private var offsetY: Float = 0f
    private var viewWidth: Int = 0
    private var viewHeight: Int = 0

    fun update(viewWidth: Int, viewHeight: Int) {
        if (viewWidth <= 0 || viewHeight <= 0) {
            return
        }

        this.viewWidth = viewWidth
        this.viewHeight = viewHeight

        val scaleX = viewWidth.toFloat() / GameConstants.WORLD_WIDTH
        val scaleY = viewHeight.toFloat() / GameConstants.WORLD_HEIGHT
        scale = minOf(scaleX, scaleY)

        val contentWidth = GameConstants.WORLD_WIDTH * scale
        val contentHeight = GameConstants.WORLD_HEIGHT * scale
        offsetX = (viewWidth - contentWidth) * 0.5f
        offsetY = (viewHeight - contentHeight) * 0.5f
    }

    fun screenToWorldX(screenX: Float): Int {
        return Math.round((screenX - offsetX) / scale)
    }

    fun screenToWorldY(screenY: Float): Int {
        return Math.round((screenY - offsetY) / scale)
    }

    fun getLetterboxDestRect(): Rect {
        val rect = Rect()
        rect.left = Math.round(offsetX)
        rect.top = Math.round(offsetY)
        rect.right = Math.round(offsetX + GameConstants.WORLD_WIDTH * scale)
        rect.bottom = Math.round(offsetY + GameConstants.WORLD_HEIGHT * scale)
        return rect
    }

    fun getScale(): Float {
        return scale
    }

    fun getOffsetX(): Float {
        return offsetX
    }

    fun getOffsetY(): Float {
        return offsetY
    }

    fun getViewWidth(): Int {
        return viewWidth
    }

    fun getViewHeight(): Int {
        return viewHeight
    }
}
